contracts/strategies/traderjoe/TraderJoeInvestmentLib.sol

```diff
@@ -0,0 +1,319 @@
+            .pairDepositToken
+            .balanceOf(address(this));
+
+        strategyStorage.lbPair.collectFees(
+            address(this),
+            strategyStorage.binIds
+        );
+
+        uint256 pairDepositTokenAfter = strategyStorage
+            .pairDepositToken
+            .balanceOf(address(this));
+
+        uint256 pairDepositTokenIncrement = pairDepositTokenAfter -
+            pairdepositTokenBefore;
+
+        swapTokens(
+            pairDepositTokenIncrement,
+            strategyStorage.pairDepositToken,
+            strategyStorage.depositToken
+        );
+    }
+
+    function adjustBins(
+        uint256[] calldata binIds,
+        uint256[] calldata binAllocations,
+        uint256 investmentTokenSupply
+    ) public {
+        TraderJoeStorage storage strategyStorage = TraderJoeStorageLib
+            .getStorage();
+
+        // Withdraw from all bins.
+        uint256 depositTokenBefore = strategyStorage.depositToken.balanceOf(
+            address(this)
+        );
+        uint256 pairDepositTokenBefore = strategyStorage
+            .pairDepositToken
+            .balanceOf(address(this));
+
+        withdraw(investmentTokenSupply, investmentTokenSupply);
+
+        uint256 depositTokenAfter = strategyStorage.depositToken.balanceOf(
+            address(this)
+        );
+        uint256 pairDepositTokenAfter = strategyStorage
+            .pairDepositToken
+            .balanceOf(address(this));
+
+        // Set bin IDs and allocations to the given ones.
+        strategyStorage.binIds = binIds;
+        strategyStorage.binAllocations = binAllocations;
+
+        // Deposit into the new bins with the new allocations.
+        uint256 depositTokenIncrement = depositTokenAfter - depositTokenBefore;
+        uint256 pairDepositTokenIncrement = pairDepositTokenAfter -
+            pairDepositTokenBefore;
+
+        deposit(depositTokenIncrement, pairDepositTokenIncrement);
+    }
+
+    function swapTokens(
+        uint256 amountIn,
+        IERC20Upgradeable tokenIn,
+        IERC20Upgradeable tokenOut
+    ) public returns (uint256 amountOut) {
+        TraderJoeStorage storage strategyStorage = TraderJoeStorageLib
+            .getStorage();
+
+        (amountOut, ) = strategyStorage.lbRouter.getSwapOut(
+            address(strategyStorage.lbPair),
+            amountIn,
+            strategyStorage.tokenY == strategyStorage.depositToken
+        );
+
+        if (amountOut == 0) {
+            return 0;
+        }
+
+        address[] memory path = new address[](2);
+        path[0] = address(tokenIn);
+        path[1] = address(tokenOut);
+
+        uint256[] memory binSteps = new uint256[](1);
+        binSteps[0] = strategyStorage.binStep;
+
+        amountOut = SwapServiceLib.swapExactTokensForTokens(
+            strategyStorage.swapService,
+            amountIn,
+            0,
+            path,
+            binSteps
+        );
+    }
+
+    function swapBack(uint256 amountIn) public {
```

Comment:
why is it not 'internal'? If you call the an internal library function from a library function, that code will go into the library's contract code.

Reply:
See commit [bb6f38f](https://github.com/block42-blockchain-company/bro-strategies/pull/34/commits/bb6f38ff3498188c475f047fbfeb2f7d203f14ab)

Reply:
'Extract investment code to library due to contract size limit' - and cannot you make this function internal and call it from the lib? That will only increase the lib's size. Public is not the right visibility here.